components/raftstore/src/store/worker/check_leader.rs

```diff
@@ -0,0 +1,166 @@
+use std::collections::Bound::{Excluded, Unbounded};
+use std::fmt;
+use std::sync::Arc;
+use std::sync::Mutex;
+use tikv_util::worker::Runnable;
+
+pub struct Runner {
+    store_meta: Arc<Mutex<StoreMeta>>,
+    region_read_progress: RegionReadProgressRegistry,
+}
+
+pub enum Task {
+    // Check if the provided `LeaderInfo`s are same as ours local `LeaderInfo`
+    CheckLeader {
+        leaders: Vec<LeaderInfo>,
+        cb: Box<dyn FnOnce(Vec<u64>) + Send>,
+    },
+    // Get the minimal `safe_ts` from regions overlap with the key range [`start_key`, `end_key`)
+    GetStoreTs {
+        key_range: KeyRange,
+        cb: Box<dyn FnOnce(u64) + Send>,
+    },
+}
+
+impl fmt::Display for Task {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut de = f.debug_struct("CheckLeaderTask");
+        match self {
+            Task::CheckLeader { ref leaders, .. } => de
+                .field("name", &"check_leader")
+                .field("leader_num", &leaders.len())
+                .finish(),
+            Task::GetStoreTs { ref key_range, .. } => de
+                .field("name", &"fet_store_ts")
+                .field("key_range", &key_range)
+                .finish(),
+        }
+    }
+}
+
+impl Runner {
+    pub fn new(store_meta: Arc<Mutex<StoreMeta>>) -> Runner {
+        let region_read_progress = store_meta.lock().unwrap().region_read_progress.clone();
+        Runner {
+            region_read_progress,
+            store_meta,
+        }
+    }
+
+    // Get the minimal `safe_ts` from regions overlap with the key range [`start_key`, `end_key`)
+    fn get_range_safe_ts(&self, key_range: KeyRange) -> u64 {
+        if key_range.get_start_key().is_empty() && key_range.get_end_key().is_empty() {
+            // Fast path to get the min `safe_ts` of all regions in this store
+            self.region_read_progress.map(|registry| {
+                registry
+                .iter()
+                .map(|(_, rrp)| rrp.safe_ts())
+                .filter(|ts| *ts != 0) // ts == 0 means the peer is uninitialized
+                .min()
+                .unwrap_or(0)
+            })
+        } else {
+            // `store_safe_ts` won't be accessed frequently (like per-request or per-transaction),
+            // also this branch won't entry because the request key range is empty currently (in v5.1)
+            // keep this branch for robustness and future use, so it is okay getting `store_safe_ts`
+            // from `store_meta` (behide a mutex)
+            let meta = self.store_meta.lock().unwrap();
+            let (start_key, end_key) = (
+                data_key(key_range.get_start_key()),
+                data_end_key(key_range.get_end_key()),
+            );
+            meta.region_read_progress.map(|registry| {
+                meta.region_ranges
+                // get overlapped regions
+                .range((Excluded(start_key.clone()), Unbounded))
```

Comment:
Why clone?

Reply:
Because `map` takes `FnMut`, I change it to `FnOnce` and remove the `clone` now.